plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -70,36 +70,46 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
+    }
 
-        return when (whenExpression.subject) {
-            null -> {
-                val cvar = data.newAnonVar(data.convertType(whenExpression.typeRef.coneTypeOrNull!!))
-                val cond = whenExpression.branches[0].condition.accept(this, data)
-                // TODO: tidy this up and split it into helper functions.
-                val thenCtx = StmtConverter(data)
-                val thenResult = whenExpression.branches[0].result.accept(this, thenCtx)
-                thenCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), thenResult))
-                val elseCtx = StmtConverter(data)
-                val elseResult = whenExpression.branches[1].result.accept(this, elseCtx)
-                elseCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), elseResult))
-                data.addDeclaration(cvar.toLocalVarDecl())
-                data.addStatement(
-                    Stmt.If(
-                        cond,
-                        thenCtx.block,
-                        elseCtx.block
-                    )
-                )
-                cvar.toLocalVar()
+    private fun visitWhenBranches(whenBranches: List<FirWhenBranch>, data: StmtConversionContext, cvar: ConvertedVar): Exp {
+        // NOTE: I think that this will also work with "in" or "is" conditions when implemented, but I'm not 100% sure
+        val cond = whenBranches[0].condition.accept(this, data)
+        val thenCtx = StmtConverter(data)
+        val thenResult = whenBranches[0].result.accept(this, thenCtx)
+        thenCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), thenResult))
+        val elseBranch =
+            when {
+                // When last branch is not an else
+                whenBranches.size == 1 -> Stmt.Seqn(listOf(), listOf())
+                // When last branch is an else
+                whenBranches.size == 2 && whenBranches[1].condition is FirElseIfTrueCondition -> {
+                    val elseCtx = StmtConverter(data)
+                    val elseResult = whenBranches[1].result.accept(this, elseCtx)
+                    elseCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), elseResult))
+                    elseCtx.block
+                }
+                // Recursive case
+                else -> {
+                    val elseCtx = StmtConverter(data)
+                    visitWhenBranches(whenBranches.drop(1), elseCtx, cvar)
+                    elseCtx.block
+                }
             }
-            else -> TODO("Can't embed $whenExpression since when expressions with a subject other than null are not yet supported.")
+        data.addStatement(Stmt.If(cond, thenCtx.block, elseBranch))
+        return cvar.toLocalVar()
+    }
+
+    override fun visitWhenExpression(whenExpression: FirWhenExpression, data: StmtConversionContext): Exp {
+        if (whenExpression.branches.isEmpty()) {
+            throw Exception("When expression $whenExpression has no branches")
```

Comment:
Is this forbidden?  Or does it just evaluate to unit?

Reply:
surprisingly this is valid code:
```
fun <!VIPER_TEXT!>when_without_branches<!>(): Int {
    when {

    }
    return 1
}
```
Do you think that it is better to evaluate to unit?
Probably I should add a test for this

Reply:
Could you check what the runtime behaviour is, and what the type of `x` is if you do `val x = when {}`?  I do think we should handle this case, even if it's niche.

Reply:
`val x = when {}` is not valid code because " 'when' expression must be exhaustive, add necessary 'else' branch".
The only way to achieve something similar that I found is having a subject of a type with no inhabitants, but is very contrived.
```
fun <!VIPER_TEXT!>when_empty<!>(n: Nothing): Int {
    val x = when(n){

    }
    return 1
}
```
In this case x has type Unit

Reply:
Let's handle it by making it a no-op, then.  Actually, I think if you make zero the base case you'll get this by default.